activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -516,6 +521,20 @@ def default_index_type?(index) # :nodoc:
 
       private
 
+        def with_connection_verify(exception_class)
+          tries = 3
+          begin
+            yield
+          rescue exception_class
+            tries -= 1
+            raise if tries <= 0 || @connection_verified
+            verify!
+            retry
```

Comment:
We need to be careful about what is being retried here.

We do NOT want to retry:
* queries that timeout
* write queries (or just only retry if the first attempt didn't even manage to send the query)
* queries inside a transaction (`@connection_verified` should avoid this, but a comment is needed to make that clear in case `@connection_verified` is removed in the future)